### PR TITLE
Test and document otel.javaagent.configuration-file

### LINKED ISSUE #2454: Add a test for otel.javaagent.configuration-file
This is a public configuration property, so we should have at least one test for it.

### PR DESCRIPTION
**Pull request by Enkelian**
Resolves #2454 and #2987

### CODE REVIEW

File: javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/config/ConfigurationFileTest.groovy [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    when:
+    def properties = ConfigInitializer.loadConfigurationFile()
+
+    then:
+    !properties.propertyNames().hasMoreElements()
+  }
+
+  def createFile(String name, String contents){
+    def path = ""
+    File.createTempFile(name, "").with {
```
Comment (mateuszrzeszutek) [inline javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/config/ConfigurationFileTest.groovy:82]: Consider using JUnit's `TemporaryFolder` rule instead.

File: docs/agent-config.md [hunk 2]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
...
+* SPI
+* hard-coded defaults
+
+### File configuration
+
+You can provide a path to agent configuration file by setting the corresponding property.
+
+| System property                      | Environment variable                 | Description                                                                      |
+|--------------------------------------|--------------------------------------|----------------------------------------------------------------------------------|
+| `otel.javaagent.configuration-file` | `OTEL_JAVAAGENT_CONFIGURATION_FILE` | Path to configuration file which contains a line-oriented list of properties formatted like `property1=value1` or `property2:value2`|
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:54]: I think it's fine to just say that it's a Java properties file without describing the format in detail:
```suggestion
| `otel.javaagent.configuration-file` | `OTEL_JAVAAGENT_CONFIGURATION_FILE` | Path to valid Java properties file which contains the javaagent configuration. |
```

File: docs/agent-config.md [hunk 3]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
 * [Span limits](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#span-limits)
 * [Using SPI to further configure the SDK](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#customizing-the-opentelemetry-sdk)
 
+## Supplying configuration properties
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:39]:
```suggestion
## Configuring the agent
```

File: docs/agent-config.md [hunk 4]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
...
+
+Configuration properties can be supplied with one or more of the following ways (priority from highest to lowest):
+* system properties
+* environment variables
+* configuration file
+* SPI
+* hard-coded defaults
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:46]: There's no need to mention these here, the user cannot do anything with those.

File: docs/agent-config.md [hunk 5]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
...
+
+### File configuration
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:48]:
```suggestion
### Configuration file
```

File: docs/agent-config.md [hunk 6]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
 * [Span limits](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#span-limits)
 * [Using SPI to further configure the SDK](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#customizing-the-opentelemetry-sdk)
 
+## Supplying configuration properties
+
+Configuration properties can be supplied with one or more of the following ways (priority from highest to lowest):
+* system properties
+* environment variables
+* configuration file
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:44]:
```suggestion
* the [configuration file](#configuration-file)
```

File: docs/agent-config.md [hunk 7]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
...
+* SPI
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:45]:
```suggestion
* the [`ConfigPropertySource`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/config/ConfigPropertySource.java) SPI
```

File: docs/agent-config.md (hunk 2 above)
Comment (iNikem) [inline docs/agent-config.md:54]: "Java Properties format" is the standard way to describe such files. I stumbled for a second on "line-oriented" :D

File: docs/agent-config.md [hunk 8]
```diff
@@ -35,6 +36,24 @@ Here are some quick links into those docs for the configuration options for spec
 * [Span limits](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#span-limits)
 * [Using SPI to further configure the SDK](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure/README.md#customizing-the-opentelemetry-sdk)
 
+## Supplying configuration properties
+
+Configuration properties can be supplied with one or more of the following ways (priority from highest to lowest):
```
Comment (mateuszrzeszutek) [inline docs/agent-config.md:41]:
```suggestion
The agent can consume configuration from one or more of the following sources (ordered from highest to lowest priority):
```

**Review by mateuszrzeszutek — COMMENTED** [review]
CC @theletterf

**Review by iNikem — APPROVED** [review]

File: docs/agent-config.md (hunk 4 above)
Comment (Enkelian) [reply docs/agent-config.md:46]: Good point, I'll get rid of it

File: docs/agent-config.md (hunk 2 above)
Comment (Enkelian) [reply docs/agent-config.md:54]: Yep, that's certainly better, thanks :)

File: javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/config/ConfigurationFileTest.groovy (hunk 1 above)
Comment (Enkelian) [reply javaagent-tooling/src/test/groovy/io/opentelemetry/javaagent/tooling/config/ConfigurationFileTest.groovy:82]: Thanks, I'll replace it

**Review by mateuszrzeszutek — APPROVED** [review]

**Review by trask — APPROVED** [review]
thx!

**Pull request merged**